### DISCUSSION TITLE
Bottom Bar Layout RFC (Request for Comments)

### DISCUSSION THREAD

**Discussion by essential-randomness** [BobaBoard/issues#98] [Design Discussions]
## The problem

People have been clamoring for more indicators and actions on threads/boards, and we don't know where to put them, especially on mobile. An easy example is "some threads lose the star jumper" (gallery/timeline mode swap it for a "add top level contrib"), but there are other open issues like "I'd like to know if the thread is muted from within it" or "I can't see if I'm logged on on mobile".

## The proposal

Change the mobile layout to have a bottom bar, and expand the functionality of the "floating action button" to be a "floating action bar".

Here are some low-fi mocks:

### Board view
![image](https://user-images.githubusercontent.com/47095486/192364100-3cf0dcdc-0a99-4359-ac0c-d519ab9f1b54.png)
![image](https://user-images.githubusercontent.com/47095486/192364142-1ce0a16c-1fe2-44d2-809f-4bae457a0317.png)

### Thread view

![image](https://user-images.githubusercontent.com/47095486/192364362-dd04b55a-9558-40a9-9cee-582780dfdcf1.png)![image](https://user-images.githubusercontent.com/47095486/192364301-919abeff-d72b-491d-9769-3afacbe147b6.png)

## Changes highlights

1. Obviously the most visible changes is **the bottom bar on mobile**. Note that I've also vertically squeezed the top bar to give us back some real estate. 
![image](https://user-images.githubusercontent.com/47095486/192366180-a62beb85-f800-44df-a4ad-4904e526c10c.png)
2. The "compass menu icon" is now swapped with the **online/login indicator on the mobile top bar**. The "bring up the contextual sidebar" action is now on the mobile bottom bar (the one with the ? button).
![image](https://user-images.githubusercontent.com/47095486/192366287-3f4432b0-4e5c-467d-a8be-62751c56686e.png)
![image](https://user-images.githubusercontent.com/47095486/192366304-d73b24c3-db12-4f36-8bbb-1b0c9b7263fc.png)
3.  The **"beam to next" indicator is always visible**, on both threads and board. When there is something new (on threads), this will change in appearance (I imagine it will get e.g. a colored red dot on top) and it will become a "beam to next _new_". Alternatively, it's always there but grayed out if there are no new things.
![image](https://user-images.githubusercontent.com/47095486/192366372-aa833201-a810-46fb-8670-30de59419d7d.png)
![image](https://user-images.githubusercontent.com/47095486/192366392-e1622baf-7996-43c5-ac65-8defab448c0c.png)
4. There is an always-available **new thread button** on the top bar of the desktop layout, which is where other socials usually put it (see Tumblr).
![image](https://user-images.githubusercontent.com/47095486/192367073-a1888d28-9cb0-4ed2-975f-693ef722a6d4.png)
5. One of the bottom bar slots has **indicators for the "current state of the board/thread you're looking at"** (e.g. pinned, hidden, muted, starred). On desktop, we move this information to the top of the sidebar (and we could consider having a shortened version of the sidebar follow us down while we scroll so the info stays in view). We will need to do a more hi-fi mock of the mobile bar to ensure these indicators look nice there and don't look like confusing buttons.
![image](https://user-images.githubusercontent.com/47095486/192369496-f3945504-ec40-4d6a-bf3b-df8368a38f55.png)
![image](https://user-images.githubusercontent.com/47095486/192369309-af77ec0d-cc02-45c6-948f-de755aa2467b.png)
6. The bottom bar actions that don't have space somewhere else on desktop, are moved to a **"floating action bar"** at the bottom of the page (note: the `...` are because I don't exactly know what other actions might go there, and _in theory_ I will want those to be configurable, _especially_ if we ever get a plugin system). When you're in a thread, the floating action bar always gets a **"add new _top level_ contribution" button**, regardless of view mode. To distinguish this from the "create new thread" button up top, we should consider changing the icon of the "new thread" action and make it different from "new contrib" so the difference is clear. On mobile, the "new thread" button is not available within thread (the central button is always "new top level contrib"). 
![image](https://user-images.githubusercontent.com/47095486/192370963-7b641e77-5237-45b8-bea0-d45258e10a51.png)
7. **Bonus** (not pictured): when we switch to the new editor, we can use the bottom bar to display the formatting options somewhere more easily tappable than within the "floating bubble" that appears on new line or when you're selecting the text.

## Nest steps

Let me know what you think, and I'll make a more hi-fi mock in (likely) Figma. If you want, you can checkout the ["bottom bar proposal" branch of boba-docs](https://github.com/BobaBoard/boba-docs/tree/bottom-bar-proposal) and edit the [`docusaurus/docs/product/mockups/proposal.excalidraw`](https://github.com/BobaBoard/boba-docs/blob/bottom-bar-proposal/docusaurus/docs/product/mockups/proposal.excalidraw) file to make changes to my mocks and play around with different things. You can use [the VSCode Excalidraw extension](https://marketplace.visualstudio.com/items?itemName=pomdtr.excalidraw-editor) to work on it directly from VSCode.

**Comment by enigmalea** [comment]
Thinking of real estate on mobile: Will the bottom bar still be visible when the keyboard appears?

If so, will that impact the editor negatively on our smallest supported screen size?

Could there be a toggle to hide it or show it somehow?

**Reply by essential-randomness** [reply]
For when the keyboard is out, I was thinking something very similar to what GitHub has given me trying to reply to this comment 😂

See:
![Screenshot_20220926-222315.png](https://user-images.githubusercontent.com/47095486/192439205-9d71cae0-312d-4ad3-a4ab-a743771c1a84.png)

If someone with a smaller screen wants to try the GitHub app and tell me how that feels, that would be helpful. Generally I think that there's a lot of screen real estate we can get back with a more thoughtful design of the editors, so we can make the change a lot less drastic than it might at first seem.

**Reply by essential-randomness** [reply]
For the toggle, I can think of a couple of interactions I've seen on other app where the bottom bar gets hidden e.g. when you tap or as you scroll, but my impression is that they're finicky, and that it'd be a lot more UX design work than it's worth to make them feel natural.

I had in the past considered a bottom bar that could hide to the side (something like a swipeable floating action button), but tbh it would still be disruptive anyway. If someone has a good proposal for an hideable bottom bar, I'll gladly consider it, but TBH bottom bars have grown on me, and we can always do more fancy hiding later on.

**Reply by enigmalea** [reply]
I agree about those things usually being more trouble than they are worth, for what it's worth. I actually noticed the github bar as I was typing on mobile, too, and was like "oh well that's nice." 

As long as it works for smaller screens I say we use something like that rather than try to overengineer something.

**Reply by essential-randomness** [reply]
I've gotten myself a screenshot of Twitter (which has a bottom bar) on the smallest phone we support with a keyboard out. There's enough space there, though I do feel like it might be good for us to revisit the mobile post layout. We're clearly wasting a lot of space and could compact things more.

![image](https://user-images.githubusercontent.com/47095486/192608282-0a18c255-4015-4ec6-8537-9e3d9d7dbcec.png)

(Our own, same device, with no keyboard cause I got it from webdev tools)
![image](https://user-images.githubusercontent.com/47095486/192608698-1b354d33-9cc4-4bec-a5cf-10d6ae9e2ced.png)

**Reply by Reptapog** [reply]
i have difficulties remembering what im replying too sometimes and many of these sorts of methods of handling replying never have an option to swipe the block out of the way or scroll up to reread the comment youre responding too so i think it would be worthwhile, if this is how it's implemented, to make sure theres an option to move the reply away somehow so the og comment or post can be viewed again without outright cancelling or throwing away the post the user is making

**Reply by riazaia** [reply]
![image](https://user-images.githubusercontent.com/42520601/193179140-851fdb63-75cb-4807-ace3-e2017d67fc08.jpeg)

edit: sorry for the big ass picture XD But there you go. I was gonna write a message but the fact that the interface says "Save" made my brain glitch (I've been conditioned by Boba herself, whose buttons say "Post"? Who knows.) 
I kinda forgot what my comment was but uh, something along the lines of Agree with @Reptapog! Having a little icon to minimize the reply would be great on mobile, especially. 

edit2: Oh! and that the size seems good to me! You can fit like 5 lines of text and that's even better than on other apps tbh.

**Comment by Reptapog** [comment]
main thought right now with the given mocks and explanations is that im looking forward to this change for mobile, its gonna make the whole experience that much better even if there'll still be things to clean up after the fact

**Reply by essential-randomness** [reply]
> i have difficulties remembering what im replying too sometimes 

Same, tbh. This is a thing I repeatedly curse my own "editor modal" implementation for.

**Comment by essential-randomness** [comment]
This has finally been implemented in https://github.com/BobaBoard/boba-frontend/pull/104!
